cooked-validators/src/Cooked/Attack/AddConstraints.hs

```diff
@@ -0,0 +1,264 @@
+--
+-- - Adding a 'SpendsPK' for an UTxO which is already consumed by the unmodified
+--   transaction will leave the transaction unmodified.
+--
+-- - Adding a 'SpendsScript' for an UTxO that is already consumed by the
+--   unmodified transaction will lead to failure, unless the 'SpendsScript'
+--   constraint on the unmodified transaction uses exactly the same validator,
+--   redeemer, and datum. In that case, the attack will leave the transaction
+--   unmodified.
+--
+-- - 'Mints' constraints are added without any additional checks.
+--
+-- - if any constraint concerning time (i.e. 'Before', 'After', or 'ValidateIn')
+--   is added, the resulting transaction will only have one 'ValidateIn'
+--   constraint, specifying as the validity time range the intersection of all
+--   the pre-existing time ranges on the transaction and the new time range.
+--
+-- - if a 'SignedBy' constraint is added, the resulting transaction will
+--   similarly have only one 'SignedBy' constraint, which holds all of the
+--   signers required by the unmodified transaction, together with the new
+--   signers.
+--
+-- The returned 'MiscConstraint' is either @Just@ the constraint that was added
+-- to the transaction (which might be different from the one you specified in
+-- case of time constraints, in which case the returned constraint holds the new
+-- validity time range of the transaction), or @Nothing@, if your constraint was
+-- already present on the unmodiified transaction.
+addMiscConstraintAttack :: MiscConstraint -> Attack (Maybe MiscConstraint)
+addMiscConstraintAttack (SpendsScript v r (o, d)) =
+  do
+    added <- addSpendsScriptAttack v r (o, d)
+    if added
+      then return $ Just $ SpendsScript v r (o, d)
+      else return Nothing
+addMiscConstraintAttack (SpendsPK o) =
+  do
+    added <- addSpendsPKAttack o
+    if added
+      then return $ Just $ SpendsPK o
+      else return Nothing
+addMiscConstraintAttack (Mints r ps x) =
+  addMintsAttack r ps x
+    >> return (Just $ Mints r ps x)
+addMiscConstraintAttack (Before b) =
+  let leftUnbounded = Pl.to b -- is this correct, i.e. should the end time be excluded?
+   in addValidateInAttack leftUnbounded
+        >>= return . Just . ValidateIn . Pl.intersection leftUnbounded
+addMiscConstraintAttack (After a) =
+  let rightUnbounded = Pl.from a -- is this correct, i.e. should the start time be excluded?
+   in addValidateInAttack rightUnbounded
+        >>= return . Just . ValidateIn . Pl.intersection rightUnbounded
+addMiscConstraintAttack (ValidateIn range) =
+  addValidateInAttack range
+    >>= return . Just . ValidateIn . Pl.intersection range
+addMiscConstraintAttack (SignedBy s) = addSignedByAttack s >>= return . Just . SignedBy
+
+-- | This attack removes some 'MiscConstraint's from a transaction. It returns a
+-- list of the removed constraints.
+removeMiscConstraintsAttack :: (MiscConstraint -> Bool) -> Attack [MiscConstraint]
```

Comment:
Another possible interface could be
```haskell
removeMiscConstraintsAttack :: (MiscConstraint -> Maybe a) -> Attack [a]
```
thus the pattern
```haskell
old <- removeMiscConstraintsAttack p
return $ summarize (map convert old)
```
becomes
```haskell
old <- removeMiscConstraintsAttack p'  -- p' c == if p c then Just (convert c) else Nothing
return (summarize old)
```
The original formulation forces `p` and `convert` to pattern match the same way on the same constraints, whereas in the later `p'` can pattern match only once.

Reply:
Just curious, what was the resolution here?

Reply:
Oh, I think it might be in the code, but I wasn't following the right links to find it.